### PR TITLE
add type argument for ngx.req.escape_uri

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

releated pr:
https://github.com/openresty/lua-resty-core/pull/303
https://github.com/openresty/lua-nginx-module/pull/1709
https://github.com/openresty/meta-lua-nginx-module/pull/77

### CODE REVIEW

File: src/ngx_http_lua_string.c [hunk 1]
```diff
@@ -433,17 +433,20 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
...
 
 size_t
-ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len)
+ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len,
+    int not_component)
 {
-    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len,
-                                             NGX_ESCAPE_URI_COMPONENT);
+    int type = not_component ? NGX_ESCAPE_URI :  NGX_ESCAPE_URI_COMPONENT;
+    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len, type);
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:440]: Let's use a `type` component just like the nginx API so that we can support `NGX_ESCAPE_ARGS` in the future as well instead of adding even more optional boolean args here.

File: README.markdown [hunk 2]
```diff
@@ -5573,6 +5573,15 @@ ngx.escape_uri
 
 Escape `str` as a URI component.
 
+Since `v0.10.16rc6`, this function accepts an optional boolean `not_component` argument. When this argument is `true`, these characters bellow will not be escaped.
```
Comment (agentzh) [inline README.markdown:5576]: I think `no_component` is better?

File: src/ngx_http_lua_string.c [hunk 3]
```diff
@@ -433,17 +433,20 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
 
 
 size_t
-ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len)
+ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len,
+    int not_component)
 {
-    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len,
-                                             NGX_ESCAPE_URI_COMPONENT);
+    int type = not_component ? NGX_ESCAPE_URI :  NGX_ESCAPE_URI_COMPONENT;
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:439]: Style: seems like you use two spaces after the colon (`:`)?

I don't think you need to introduce the `type` variable in the first place.

File: src/ngx_http_lua_string.c [hunk 4]
```diff
@@ -433,17 +433,20 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
...
 }
 
 
 void
-ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst)
+ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst,
+    int not_component)
 {
-    ngx_http_lua_escape_uri(dst, (u_char *) src, len, NGX_ESCAPE_URI_COMPONENT);
+    int type = not_component ? NGX_ESCAPE_URI :  NGX_ESCAPE_URI_COMPONENT;
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:448]: Ditto.

File: doc/HttpLuaModule.wiki [hunk 5]
```diff
@@ -4678,6 +4678,15 @@ This method was introduced in the <code>0.5.0rc30</code> release.
...
 
+Since `v0.10.16rc6`, this function accepts an optional boolean <code>not_component</code> argument. When this argument is <code>true</code>, these characters bellow will not be escaped.
+
+<geshi lang="text">
+Alphabets: a-zA-Z
+Digits: 0-9
+Reserve characters: -_.~!*'();:@&=+$,/?#
+</geshi>
+
+
```
Comment (agentzh) [inline doc/HttpLuaModule.wiki:4689]: Extra blank line here?

File: README.markdown (hunk 2 above)
Comment (agentzh) [reply README.markdown:5576]: It's still "not_component"?

Comment (agentzh) [reply README.markdown:5576]: I've thought more about this. Mabye renaming the argument to `not_uri_comp` is a better.

File: README.markdown [hunk 6]
```diff
@@ -5573,6 +5573,15 @@ ngx.escape_uri
...
+
+
+    Alphabets: a-zA-Z
+    Digits: 0-9
+    Reserve characters: -_.~!*'();:@&=+$,/?#
+
+
+
```
Comment (agentzh) [inline README.markdown:5584]: Extra blank lines here.

File: src/ngx_http_lua_string.c [hunk 7]
```diff
@@ -433,17 +433,20 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
 
 
 size_t
-ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len)
+ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len,
+    int not_component)
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:437]: Ditto. Better be `not_uri_comp`.

File: README.markdown [hunk 8]
```diff
@@ -5567,12 +5567,19 @@ This method was introduced in the `0.5.0rc30` release.
...
 --------------
 
-**syntax:** *newstr = ngx.escape_uri(str)*
+**syntax:** *newstr = ngx.escape_uri(str, not_uri_comp?)*
 
 **context:** *init_by_lua&#42;, init_worker_by_lua&#42;, set_by_lua&#42;, rewrite_by_lua&#42;, access_by_lua&#42;, content_by_lua&#42;, header_filter_by_lua&#42;, body_filter_by_lua&#42;, log_by_lua&#42;, ngx.timer.&#42;, balancer_by_lua&#42;, ssl_certificate_by_lua&#42;, ssl_session_fetch_by_lua&#42;, ssl_session_store_by_lua&#42;*
 
 Escape `str` as a URI component.
 
+Since `v0.10.16rc6`, this function accepts an optional boolean `not_uri_comp` argument. When this argument is `true`, these characters bellow will be escaped.
```
Comment (spacewander) [inline README.markdown:5576]: Personally speaking, it would be more extendable to use an escape type enumeration than to use a boolean variable. 

Maybe we can also list the escaped  characters range when escaping `str` as a URI component as comparison?

File: src/ngx_http_lua_consts.c [hunk 9]
```diff
@@ -13,6 +13,34 @@
...
+    lua_setfield(L, -2, "NGX_ESCAPE_HTML");
+
+    lua_pushinteger(L, NGX_ESCAPE_REFRESH);
+    lua_setfield(L, -2, "NGX_ESCAPE_REFRESH");
+
+    lua_pushinteger(L, NGX_ESCAPE_MEMCACHED);
+    lua_setfield(L, -2, "NGX_ESCAPE_MEMCACHED");
+
+    lua_pushinteger(L, NGX_ESCAPE_MAIL_AUTH);
+    lua_setfield(L, -2, "ERNGX_ESCAPE_MAIL_AUTHROR");
```
Comment (spacewander) [inline src/ngx_http_lua_consts.c:39]: Is the `ERNGX` intended?

File: README.markdown (hunk 8 above)
Comment (agentzh) [reply README.markdown:5576]: Yeah, that was also what I original proposed. The optional `type` argument can be of the integer type. This way we can also support other types like `NGX_ESCAPE_ARGS`, `NGX_ESCAPE_HTML`, and etc, just like nginx's own `ngx_escape_uri()` C function.

File: src/ngx_http_lua_consts.c (hunk 9 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_consts.c:39]: sorry

File: README.markdown [hunk 10]
```diff
@@ -3592,6 +3593,27 @@ These constants are usually used by the [ngx.log](#ngxlog) method.
...
+
+```lua
+
+   ngx.ESCAPE_URI
+   ngx.ESCAPE_ARGS
+   ngx.ESCAPE_URI_COMPONENT
+   ngx.ESCAPE_HTML
+   ngx.ESCAPE_REFRESH
+   ngx.ESCAPE_MEMCACHED
+   ngx.ESCAPE_MAIL_AUTH
```
Comment (agentzh) [inline README.markdown:3609]: Nowadays we try to avoid bloating the `ngx` table. Better avoid adding these fields. Also we don't have documentation to explain what these types really mean. Maybe we only need to document `URI` and `URI_COMPONENT` types for now.

File: doc/HttpLuaModule.wiki [hunk 11]
```diff
@@ -2925,6 +2925,23 @@ These constants are usually used in [[#ngx.location.capture|ngx.location.capture
...
+  ngx.ESCAPE_URI
+  ngx.ESCAPE_ARGS
+  ngx.ESCAPE_URI_COMPONENT
+  ngx.ESCAPE_HTML
+  ngx.ESCAPE_REFRESH
+  ngx.ESCAPE_MEMCACHED
+  ngx.ESCAPE_MAIL_AUTH
+</geshi>
+
+
```
Comment (agentzh) [inline doc/HttpLuaModule.wiki:2942]: Extra blank line here? I think we talked about this before.

File: doc/HttpLuaModule.wiki [hunk 12]
```diff
@@ -4672,12 +4689,15 @@ This method was introduced in the <code>0.5.0rc30</code> release.
...
 
-'''syntax:''' ''newstr = ngx.escape_uri(str)''
+'''syntax:''' ''newstr = ngx.escape_uri(str, type?)''
 
 '''context:''' ''init_by_lua*, init_worker_by_lua*, set_by_lua*, rewrite_by_lua*, access_by_lua*, content_by_lua*, header_filter_by_lua*, body_filter_by_lua*, log_by_lua*, ngx.timer.*, balancer_by_lua*, ssl_certificate_by_lua*, ssl_session_fetch_by_lua*, ssl_session_store_by_lua*''
 
 Escape <code>str</code> as a URI component.
 
+Since `v0.10.16rc6`, this function accepts an optional <code>type</code> argument.
+Note that this method accepts all [[#Nginx escape type constants|Nginx escape type constants]] as input.
```
Comment (agentzh) [inline doc/HttpLuaModule.wiki:4699]: We can simply document the magical number 1, 2, and etc here, with more explanation on what these types really mean.

File: src/ngx_http_lua_consts.c [hunk 13]
```diff
@@ -13,6 +13,34 @@
 
 #include "ngx_http_lua_consts.h"
```
Comment (agentzh) [inline src/ngx_http_lua_consts.c:15]: Forgot to use 2 blank lines here again?

File: src/ngx_http_lua_string.c [hunk 14]
```diff
@@ -433,17 +433,18 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
 
 
 size_t
-ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len)
+ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len,
+    int type)
 {
-    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len,
-                                             NGX_ESCAPE_URI_COMPONENT);
+    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len, type);
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:439]: We should validate if the type is out of range...it could be any integer value anyway.

File: src/ngx_http_lua_util.c [hunk 15]
```diff
@@ -1892,13 +1893,13 @@ ngx_http_lua_escape_uri(u_char *dst, u_char *src, size_t size, ngx_uint_t type)
         0xffffffff, /* 1111 1111 1111 1111  1111 1111 1111 1111 */
 
                     /* ?>=< ;:98 7654 3210  /.-, +*)( '&%$ #"!  */
-        0xfc00886d, /* 1111 1100 0000 0000  1000 1000 0110 1101 */
+        0x80000029, /* 1000 0000 0000 0000  0000 0000 0010 1001 */
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:1896]: Do we need to update the comment on line 1890 above accordingly? Also, why this change? Is it for sync'ing upstream nginx core changes?

File: t/006-escape.t [hunk 16]
```diff
@@ -197,3 +197,100 @@ GET /lua
...
+    location /lua {
+        content_by_lua_block {
+            ngx.say(ngx.escape_uri("https://www.google.com", 100))
+        }
+    }
+--- request
+GET /lua
+--- error_code: 500
+--- error_log eval
+qr/\[error\] \d+#\d+: \*\d+ lua entry thread aborted: runtime error: "esc_type" 100 out of range/
```
Comment (agentzh) [inline t/006-escape.t:281]: The "esc_type" thing makes no sense to the end user. We should make it more user-friendly.

File: src/ngx_http_lua_consts.c [hunk 17]
```diff
@@ -13,6 +13,34 @@
...
+    lua_setfield(L, -2, "ESCAPE_HTML");
+
+    lua_pushinteger(L, NGX_ESCAPE_REFRESH);
+    lua_setfield(L, -2, "ESCAPE_REFRESH");
+
+    lua_pushinteger(L, NGX_ESCAPE_MEMCACHED);
+    lua_setfield(L, -2, "ESCAPE_MEMCACHED");
+
+    lua_pushinteger(L, NGX_ESCAPE_MAIL_AUTH);
+    lua_setfield(L, -2, "ESCAPE_MAIL_AUTH");
```
Comment (agentzh) [inline src/ngx_http_lua_consts.c:39]: We should avoid these.

File: src/ngx_http_lua_util.c (hunk 15 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_util.c:1895]: yes

File: src/ngx_http_lua_string.c [hunk 18]
```diff
@@ -433,17 +433,24 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
...
 size_t
-ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len)
+ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len,
+    int type)
 {
-    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len,
-                                             NGX_ESCAPE_URI_COMPONENT);
+    if (type < NGX_ESCAPE_URI || type > NGX_ESCAPE_MAIL_AUTH) {
+        return -1;
+    }
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:441]: Always add a blank line after `if` code block. We've run into such things for a lot of times already...

File: src/ngx_http_lua_string.c [hunk 19]
```diff
@@ -433,17 +433,24 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
...
 
 void
-ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst)
+ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst,
+    int type)
 {
-    ngx_http_lua_escape_uri(dst, (u_char *) src, len, NGX_ESCAPE_URI_COMPONENT);
+    if (type < NGX_ESCAPE_URI || type > NGX_ESCAPE_MAIL_AUTH) {
+        return;
+    }
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:452]: Ditto.

File: doc/HttpLuaModule.wiki [hunk 20]
```diff
@@ -4686,12 +4686,20 @@ This method was introduced in the <code>0.5.0rc30</code> release.
...
 
 '''context:''' ''init_by_lua*, init_worker_by_lua*, set_by_lua*, rewrite_by_lua*, access_by_lua*, content_by_lua*, header_filter_by_lua*, body_filter_by_lua*, log_by_lua*, ngx.timer.*, balancer_by_lua*, ssl_certificate_by_lua*, ssl_session_fetch_by_lua*, ssl_session_store_by_lua*''
 
 Escape <code>str</code> as a URI component.
 
+Since `v0.10.16rc6`, this function accepts an optional <code>type</code> argument.
+<geshi lang="lua">
+  type = 0 (Escape <code>str</code> as a URI)
+  type = 2 (Escape <code>str</code> as a URI component)
+</geshi>
```
Comment (agentzh) [inline doc/HttpLuaModule.wiki:4699]: Needs a blank line before `<geshi>` *and* after `</geshi>`.

File: README.markdown [hunk 21]
```diff
@@ -5581,12 +5581,21 @@ This method was introduced in the `0.5.0rc30` release.
...
 
 Escape `str` as a URI component.
 
+Since `v0.10.16rc6`, this function accepts an optional `type` argument.
+```lua
+
+   type = 0 (Escape <code>str</code> as a URI)
+   type = 2 (Escape <code>str</code> as a URI component)
+```
+when type set to 0, " ", "#", "%", "?", 0x00-0x1F, 0x7F-0xFF will be escaped.
```
Comment (agentzh) [inline README.markdown:5596]: "when type set to 0" is not well formed English. You should say "when `type` is set to 0".

File: README.markdown [hunk 22]
```diff
@@ -5581,12 +5581,21 @@ This method was introduced in the `0.5.0rc30` release.
...
 
 **context:** *init_by_lua&#42;, init_worker_by_lua&#42;, set_by_lua&#42;, rewrite_by_lua&#42;, access_by_lua&#42;, content_by_lua&#42;, header_filter_by_lua&#42;, body_filter_by_lua&#42;, log_by_lua&#42;, ngx.timer.&#42;, balancer_by_lua&#42;, ssl_certificate_by_lua&#42;, ssl_session_fetch_by_lua&#42;, ssl_session_store_by_lua&#42;*
 
 Escape `str` as a URI component.
 
+Since `v0.10.16rc6`, this function accepts an optional `type` argument.
+```lua
+
+   type = 0 (Escape <code>str</code> as a URI)
+   type = 2 (Escape <code>str</code> as a URI component)
```
Comment (agentzh) [inline README.markdown:5594]: You cannot use `<code>` inside a code block. It does not make sense. Maybe you should use markdown's bullet list here instead.

File: README.markdown [hunk 23]
```diff
@@ -5581,12 +5581,21 @@ This method was introduced in the `0.5.0rc30` release.
...
+```
+when type set to 0, " ", "#", "%", "?", 0x00-0x1F, 0x7F-0xFF will be escaped.
+when type set to 2, all characters excepter ALPHA, DIGIT, "-", ".", "_", "~" will be escaped.
+
```
Comment (agentzh) [inline README.markdown:5598]: And you haven't mentioned what `type`'s default value is.

File: doc/HttpLuaModule.wiki [hunk 24]
```diff
@@ -4686,12 +4686,20 @@ This method was introduced in the <code>0.5.0rc30</code> release.
...
+when type set to 0, " ", "#", "%", "?", 0x00-0x1F, 0x7F-0xFF will be escaped.
+when type set to 2, all characters excepter ALPHA, DIGIT, "-", ".", "_", "~" will be escaped.
```
Comment (agentzh) [inline doc/HttpLuaModule.wiki:4701]: You should combine these two lists into a single list. Each case is explained twice above.

File: src/ngx_http_lua_string.c [hunk 25]
```diff
@@ -433,17 +433,24 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
...
 }
 
 
 void
-ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst)
+ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst,
+    int type)
 {
-    ngx_http_lua_escape_uri(dst, (u_char *) src, len, NGX_ESCAPE_URI_COMPONENT);
+    if (type < NGX_ESCAPE_URI || type > NGX_ESCAPE_MAIL_AUTH) {
```
Comment (agentzh) [inline src/ngx_http_lua_string.c:450]: I think we only need to check it in the length function. Otherwise we are checking twice.

**Comment by agentzh** [comment]
Merged.
